Thanks for this! It will be great to have this functionality. However before I can merge this, there are a few things to be addressed:

1) I updated the example application in the meantime to use the latest version of the toolbox. Can you please rebase this pull request onto the latest master?
2) There should be a separate `dev` profile under the cljsbuild profiles for this functionality. The `release` profile must use the advanced optimization because I deploy that as a demo on a server, where the resulting JavaScript needs to be as small as possible.
3) I get an error when running `lein figwheel`:
`$ lein figwheel
Figwheel: Starting server at http://localhost:3449
Focusing on build ids: release
Compiling "resources/public/js/build/example.js" from ["src/cljs" "env/dev/cljs"]...
Compiling "resources/public/js/build/example.js" failed.
clojure.lang.ExceptionInfo: failed compiling file:target/figwheel_temp/release/figwheel/connect.cljs at clojure.core$ex_info.invoke (core.clj:4591)
Caused by: clojure.lang.ExceptionInfo: No such namespace: figwheel-watch-and-reload.dev, could not locate figwheel_watch_and_reload/dev.cljs, figwheel_watch_and_reload/dev.cljc, or Closure namespace "figwheel-watch-and-reload.dev" at line 1 target/figwheel_temp/release/figwheel/connect.cljs at clojure.core$ex_info.invoke (core.clj:4591)`
This error does not seem to affect anything afterwards and the application reloads properly, but I still find this error irritating. Would you mind looking into this?

How would you use a `defonce` within the library context to preserve state? Figwheel would be very useful even without, but I'm curious about your ideas there.

Cool stuff, thanks for your work on this!
Cheers, Matthias

--- a/examples/trailing-mouse-pointer/env/dev/cljs/example/dev.cljs
+++ b/examples/trailing-mouse-pointer/env/dev/cljs/example/dev.cljs
@@ -1,0 +1,14 @@
+(ns ^:figwheel-no-load example.dev
+  (:require [example.core :as c]
+            [figwheel.client :as figwheel :include-macros true]))
+
+(enable-console-print!)
+
+(defn jscb [] 
+  (c/init))
+
+(figwheel/watch-and-reload
+  :websocket-url "ws://localhost:3449/figwheel-ws"
+  :jsload-callback jscb)
+
+(c/init)

--- a/examples/trailing-mouse-pointer/project.clj
+++ b/examples/trailing-mouse-pointer/project.clj
@@ -15,6 +15,7 @@
                  [reagent "0.5.0"]
                  [incanter "1.5.6"]
                  [clj-time "0.9.0"]
+                 [lein-figwheel "0.3.5"]
                  [org.clojure/clojurescript "0.0-3308"]]
 
   :source-paths ["src/clj/"]
@@ -24,12 +25,21 @@
   :main example.core
 
   :plugins [[lein-cljsbuild "1.0.6"]
+            [lein-figwheel "0.3.5"]
             [codox "0.8.10"]]
 
-  :cljsbuild {:builds [{:id "release"
-                        :source-paths ["src/cljs"]
-                        :compiler {:optimizations :advanced
-                                   ;:optimizations :simple
-                                   ;:source-map "resources/public/js/build/example.js.map"
+  :cljsbuild {:builds [{:id "dev"
+                        :source-paths ["src/cljs" "env/dev/cljs"]
+                        :figwheel true
+                        :compiler {:main "example.dev"
+                                   :asset-path "js/build"
+                                   :optimizations :none
                                    :output-dir "resources/public/js/build/"
-                                   :output-to "resources/public/js/build/example.js"}}]})
+                                   :output-to "resources/public/js/build/example.js"}}
+                       {:id "release"
+                        :source-paths ["src/cljs"]
+                        :figwheel true
+                        :compiler {:main "example.core"
+                                   :asset-path "js/build"
+                                   :output-to "resources/public/js/build/example.js"
+                                   :optimizations :advanced}}]})

--- a/examples/trailing-mouse-pointer/src/cljs/example/core.cljs
+++ b/examples/trailing-mouse-pointer/src/cljs/example/core.cljs
@@ -12,23 +12,27 @@
 
 (def switchboard (sb/component :client/switchboard))
 
-(sb/send-mult-cmd
-  switchboard   ;; Below, we're interacting with the switchboard component we just created above.
-  [;; First of all, we instantiate and wire a couple fo different components.
-   [:cmd/wire-comp (sente/component :client/ws-cmp)]        ; WebSocket communication component
-   [:cmd/wire-comp (hist/component :client/histogram-cmp)]  ; UI component for histograms
-   [:cmd/wire-comp (mouse/component :client/mouse-cmp)]     ; UI component for capturing mouse moves
-   [:cmd/wire-comp (store/component :client/store-cmp)]     ; Data store component
-   [:cmd/wire-comp (jvmstats/component :client/jvmstats-cmp "jvm-stats-frame")] ;  UI component: JVM stats
-   [:cmd/wire-comp (obs/component :client/observer-cmp conf/observer-cfg-map)]  ; UI component for observing system
+(defn init
+  []
+  (sb/send-mult-cmd
+    switchboard   ;; Below, we're interacting with the switchboard component we just created above.
+    [;; First of all, we instantiate and wire a couple fo different components.
+     [:cmd/wire-comp (sente/component :client/ws-cmp)]        ; WebSocket communication component
+     [:cmd/wire-comp (hist/component :client/histogram-cmp)]  ; UI component for histograms
+     [:cmd/wire-comp (mouse/component :client/mouse-cmp)]     ; UI component for capturing mouse moves
+     [:cmd/wire-comp (store/component :client/store-cmp)]     ; Data store component
+     [:cmd/wire-comp (jvmstats/component :client/jvmstats-cmp "jvm-stats-frame")] ;  UI component: JVM stats
+     [:cmd/wire-comp (obs/component :client/observer-cmp conf/observer-cfg-map)]  ; UI component for observing system
 
-   ;; Then, messages of a given type are wired from one component to another.
-   ;[:cmd/route {:from :client/mouse-cmp :to :client/ws-cmp :only :cmd/mouse-pos}]
-   [:cmd/route-all {:from :client/mouse-cmp :to :client/ws-cmp}]
-   [:cmd/route {:from :client/ws-cmp :to :client/store-cmp}]
-   [:cmd/route {:from :client/ws-cmp :to :client/jvmstats-cmp}]
-   [:cmd/observe-state {:from :client/store-cmp :to :client/histogram-cmp}]
-   [:cmd/observe-state {:from :client/store-cmp :to :client/mouse-cmp}]
+     ;; Then, messages of a given type are wired from one component to another.
+     ;[:cmd/route {:from :client/mouse-cmp :to :client/ws-cmp :only :cmd/mouse-pos}]
+     [:cmd/route-all {:from :client/mouse-cmp :to :client/ws-cmp}]
+     [:cmd/route {:from :client/ws-cmp :to :client/store-cmp}]
+     [:cmd/route {:from :client/ws-cmp :to :client/jvmstats-cmp}]
+     [:cmd/observe-state {:from :client/store-cmp :to :client/histogram-cmp}]
+     [:cmd/observe-state {:from :client/store-cmp :to :client/mouse-cmp}]
 
-   ;; Finally, wire firehose with all messages into the observer component.
-   [:cmd/attach-to-firehose :client/observer-cmp]])
+     ;; Finally, wire firehose with all messages into the observer component.
+     [:cmd/attach-to-firehose :client/observer-cmp]])
+  )
+(init)

